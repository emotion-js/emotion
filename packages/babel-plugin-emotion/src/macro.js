// @flow
import { replaceCssWithCallExpression } from './index'
import { buildMacroRuntimeNode, addRuntimeImports } from './babel-utils'
import { createMacro } from 'babel-plugin-macros'

module.exports = createMacro(macro)

function macro({ references, state, babel: { types: t } }) {
  Object.keys(references).forEach(referenceKey => {
    let isPure = true
    switch (referenceKey) {
      case 'injectGlobal': {
        isPure = false
      }
      // eslint-disable-next-line no-fallthrough
      case 'css':
      case 'keyframes': {
        references[referenceKey].reverse().forEach(reference => {
          const path = reference.parentPath
          const runtimeNode = buildMacroRuntimeNode(
            reference,
            state,
            referenceKey,
            t
          )
          if (t.isTaggedTemplateExpression(path)) {
            replaceCssWithCallExpression(
              path,
              runtimeNode,
              state,
              t,
              undefined,
              !isPure
            )
          } else {
            if (isPure) {
              path.addComment('leading', '#__PURE__')
            }
            reference.replaceWith(runtimeNode)
          }
        })
        break
      }
      default: {
        references[referenceKey].reverse().forEach(reference => {
          reference.replaceWith(
            buildMacroRuntimeNode(reference, state, referenceKey, t)
          )
        })
      }
    }
  })
  addRuntimeImports(state, t)
}
