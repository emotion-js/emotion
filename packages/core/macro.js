module.exports = require('babel-plugin-emotion').macros.core
