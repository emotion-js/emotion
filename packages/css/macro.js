module.exports = require('babel-plugin-emotion').macros.css
