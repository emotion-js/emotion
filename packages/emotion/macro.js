module.exports = require('babel-plugin-emotion').macros.emotion
