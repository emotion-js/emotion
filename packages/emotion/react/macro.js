module.exports = require('react-emotion/macro')
