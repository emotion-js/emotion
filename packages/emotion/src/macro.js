// This is only for tests
module.exports = require('babel-plugin-emotion/src/macro')
