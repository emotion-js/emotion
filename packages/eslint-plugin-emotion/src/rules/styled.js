module.exports = {
  meta: {
    fixable: 'code'
  },
  create(context) {
    return {
      ImportDeclaration(node) {
        if (
          node.source.value === 'react-emotion' ||
          node.source.value === 'preact-emotion'
        ) {
          let newImportPath =
            node.source.value === 'react-emotion'
              ? '@emotion/styled'
              : '@emotion/preact-styled'
          let quote = node.source.raw.charAt(0)
          context.report({
            node: node.source,
            message: `styled should be imported from ${newImportPath}`,
            fix:
              node.specifiers.length === 1 &&
              node.specifiers[0].type === 'ImportDefaultSpecifier'
                ? fixer => {
                    return fixer.replaceText(
                      node.source,
                      quote + newImportPath + quote
                    )
                  }
                : undefined
          })
        }
      }
    }
  }
}
