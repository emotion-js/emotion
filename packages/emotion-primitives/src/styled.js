// @flow
import * as React from 'react'
import { View, Text, Image, StyleSheet } from 'react-primitives'
import {
  channel,
  contextTypes,
  setTheme,
  testAlwaysTrue,
  pickAssign
} from './utils'
import { getStyles } from './getStyles'
import { convertToRNStyles } from './convertToRNStyles'
import {
  testPickPropsOnPrimitiveComponent,
  testPickPropsOnOtherComponent
} from './test-props'

function isPrimitiveComponent(component: React.ElementType) {
  switch (component) {
    case View:
    case Text:
    case Image: {
      return true
    }
  }
  return false
}

// Evaluate the styles and convert them to React Native using styled component context
function evalStyles(context, Comp, styles) {
  // Convert styles with or without interpolations to React Native (StyleSheet.create)
  let otherStyles = convertToRNStyles.call(context, styles)

  return getStyles.call(context, otherStyles, context.props)
}

type State = {
  theme: Object
} | null

/**
 * a function that returns a styled component which render styles on multiple targets with same code
 */
export function createStyled(component: React.ElementType) {
  let isPrimitive = isPrimitiveComponent(component)
  let pick = isPrimitive
    ? testPickPropsOnPrimitiveComponent
    : testPickPropsOnOtherComponent
  return function createStyledComponent(...styles: *) {
    class Styled extends React.Component<*, State> {
      unsubscribe: number | void
      mergedProps: Object

      static withComponent = (newComponent: React.ElementType) =>
        createStyled(newComponent)(...styles)

      componentWillMount() {
        if (this.context[channel] !== undefined) {
          this.unsubscribe = this.context[channel].subscribe(
            setTheme.bind(this)
          )
        }
      }

      componentWillUnmount() {
        if (this.unsubscribe !== undefined) {
          this.context[channel].unsubscribe(this.unsubscribe)
        }
      }

      render() {
        const { props, state } = this

        // Similar to create-emotion-styled component implementation
        this.mergedProps = pickAssign(testAlwaysTrue, {}, props, {
          theme: (state !== null && state.theme) || props.theme || {}
        })

        const emotionStyles = evalStyles(this, Styled, styles)

        return React.createElement(
          component,
          pickAssign(pick, {}, props, {
            ref: props.innerRef,
            style: StyleSheet.flatten(emotionStyles)
          })
        )
      }
    }

    Styled.contextTypes = contextTypes

    Styled.displayName = `emotion(${getDisplayName(component)})`

    return Styled
  }
}

const getDisplayName = primitive =>
  typeof primitive === 'string'
    ? primitive
    : primitive.displayName || primitive.name || 'Styled'
