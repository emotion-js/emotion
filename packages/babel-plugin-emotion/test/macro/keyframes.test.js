import React from 'react'
import renderer from 'react-test-renderer'
import { keyframes, sheet, flush } from 'emotion/macro'
import styled from 'react-emotion/macro'

describe('keyframes - macro', () => {
  afterEach(() => {
    flush()
  })
  test('renders', () => {
    const bounce = keyframes`
      from, 20%, 53%, 80%, to {
        animation-timing-function: cubic-bezier(0.215, 0.610, 0.355, 1.000);
        transform: translate3d(0,0,0);
      }
    
      40%, 43% {
        animation-timing-function: cubic-bezier(0.755, 0.050, 0.855, 0.060);
        transform: translate3d(0, -30px, 0);
      }
    
      70% {
        animation-timing-function: cubic-bezier(0.755, 0.050, 0.855, 0.060);
        transform: translate3d(0, -15px, 0);
      }
    
      90% {
        transform: translate3d(0,-4px,0);
      }
    `

    const H1 = styled.h1`animation: ${bounce} 2s linear infinite;`

    const tree = renderer.create(<H1>hello world</H1>).toJSON()

    expect(tree).toMatchSnapshot()
    expect(sheet).toMatchSnapshot()
  })
  test('keyframes with interpolation', () => {
    const endingRotation = '360deg'

    const H1 = styled.h1`
      animation: ${keyframes`
      from {
        transform: rotate(0deg);
      }
    
      to {
        transform: rotate(${endingRotation});
      }
    `} 2s linear infinite;
    `

    const tree = renderer.create(<H1>hello world</H1>).toJSON()

    expect(tree).toMatchSnapshot()

    expect(sheet).toMatchSnapshot()
  })
})
