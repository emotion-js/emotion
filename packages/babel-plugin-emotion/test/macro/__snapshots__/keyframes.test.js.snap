// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`keyframes - macro keyframes with interpolation 1`] = `
.glamor-0 {
  -webkit-animation: animation-1q8eu9e 2s linear infinite;
  animation: animation-1q8eu9e 2s linear infinite;
}

<h1
  className="glamor-0"
>
  hello world
</h1>
`;

exports[`keyframes - macro keyframes with interpolation 2`] = `
"@-webkit-keyframes animation-1q8eu9e {
  from {
    -webkit-transform: rotate(0deg);
    -ms-transform: rotate(0deg);
    transform: rotate(0deg);
  }

  to {
    -webkit-transform: rotate(360deg);
    -ms-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}

@keyframes animation-1q8eu9e {
  from {
    -webkit-transform: rotate(0deg);
    -ms-transform: rotate(0deg);
    transform: rotate(0deg);
  }

  to {
    -webkit-transform: rotate(360deg);
    -ms-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}

.css-l8b5qg {
  -webkit-animation: animation-1q8eu9e 2s linear infinite;
  animation: animation-1q8eu9e 2s linear infinite;
}"
`;

exports[`keyframes - macro renders 1`] = `
.glamor-0 {
  -webkit-animation: animation-16qlhaj 2s linear infinite;
  animation: animation-16qlhaj 2s linear infinite;
}

<h1
  className="glamor-0"
>
  hello world
</h1>
`;

exports[`keyframes - macro renders 2`] = `
"@-webkit-keyframes animation-16qlhaj {
  from, 20%, 53%, 80%, to {
    -webkit-animation-timing-function: cubic-bezier(0.215,0.610,0.355,1.000);
    animation-timing-function: cubic-bezier(0.215,0.610,0.355,1.000);
    -webkit-transform: translate3d(0,0,0);
    -ms-transform: translate3d(0,0,0);
    transform: translate3d(0,0,0);
  }

  40%, 43% {
    -webkit-animation-timing-function: cubic-bezier(0.755,0.050,0.855,0.060);
    animation-timing-function: cubic-bezier(0.755,0.050,0.855,0.060);
    -webkit-transform: translate3d(0,-30px,0);
    -ms-transform: translate3d(0,-30px,0);
    transform: translate3d(0,-30px,0);
  }

  70% {
    -webkit-animation-timing-function: cubic-bezier(0.755,0.050,0.855,0.060);
    animation-timing-function: cubic-bezier(0.755,0.050,0.855,0.060);
    -webkit-transform: translate3d(0,-15px,0);
    -ms-transform: translate3d(0,-15px,0);
    transform: translate3d(0,-15px,0);
  }

  90% {
    -webkit-transform: translate3d(0,-4px,0);
    -ms-transform: translate3d(0,-4px,0);
    transform: translate3d(0,-4px,0);
  }
}

@keyframes animation-16qlhaj {
  from, 20%, 53%, 80%, to {
    -webkit-animation-timing-function: cubic-bezier(0.215,0.610,0.355,1.000);
    animation-timing-function: cubic-bezier(0.215,0.610,0.355,1.000);
    -webkit-transform: translate3d(0,0,0);
    -ms-transform: translate3d(0,0,0);
    transform: translate3d(0,0,0);
  }

  40%, 43% {
    -webkit-animation-timing-function: cubic-bezier(0.755,0.050,0.855,0.060);
    animation-timing-function: cubic-bezier(0.755,0.050,0.855,0.060);
    -webkit-transform: translate3d(0,-30px,0);
    -ms-transform: translate3d(0,-30px,0);
    transform: translate3d(0,-30px,0);
  }

  70% {
    -webkit-animation-timing-function: cubic-bezier(0.755,0.050,0.855,0.060);
    animation-timing-function: cubic-bezier(0.755,0.050,0.855,0.060);
    -webkit-transform: translate3d(0,-15px,0);
    -ms-transform: translate3d(0,-15px,0);
    transform: translate3d(0,-15px,0);
  }

  90% {
    -webkit-transform: translate3d(0,-4px,0);
    -ms-transform: translate3d(0,-4px,0);
    transform: translate3d(0,-4px,0);
  }
}

.css-1qpx6jr {
  -webkit-animation: animation-16qlhaj 2s linear infinite;
  animation: animation-16qlhaj 2s linear infinite;
}"
`;
